lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -1482,14 +1482,21 @@ bool HandshakeBuilder::visitHandshake(MemoryOp op) {
   }
 
   // Create the special type to represent this memory.
-  FIRRTLType memType = MemOp::getTypeForPortList(depth, dataType, ports);
+  BundleType memType = MemOp::getTypeForPortList(depth, dataType, ports);
+
+  llvm::SmallVector<Type> resultTypes;
+  llvm::SmallVector<Attribute> resultNames;
+  for (auto element : memType.getElements()) {
+    resultTypes.push_back(element.type);
+    resultNames.push_back(rewriter.getStringAttr(element.name.str()));
+  }
 
   // Create the actual mem op.
-  auto memOp = rewriter.create<MemOp>(insertLoc, memType, readLatency,
-                                      writeLatency, depth, ruw, name);
+  auto memOp = rewriter.create<MemOp>(insertLoc, resultTypes, readLatency,
+                                      writeLatency, depth, ruw,
+                                      rewriter.getArrayAttr(resultNames), name);
```

Comment:
My gut is telling me that we should delete `MemOp::getTypeForPortList`, since it is a little non-sensical.  If we still need something like this, we should return an array instead of a bundle.

I think for this code, we should implement a builder that takes a portlist directly.

Reply:
I see that InstanceOp doesn't take a list of ports either.. maybe we could do something like this?

```c++
llvm::SmallVector<Type> resultTypes;
llvm::SmallVector<Attribute> resultNames;
for (auto p : ports) {
  resultTypes.push_back(FlipType::get(getTypeForPort(depth, dataType, p.second)));
  resultNames.push_back(rewriter.getStringAttr(p.first));
}

auto memOp = rewriter.create<MemOp>(insertLoc, resultTypes, readLatency,
                                    writeLatency, depth, ruw,
                                    rewriter.getArrayAttr(resultNames), name);
```

Reply:
I agree with this. It's weird to get a `BundleType` from `MemOp::getTypeForPortList` only to destructure it. With the new `getTypeForPort`, the modification is pretty clean.

Reply:
The current version uses your approach here. I'll plan to rip out `MemOp::getTypeForPortList` after this PR lands.

The one complication is that `MemOp::getTypeForPortList` is inappropriately doing canonicalization of port names into alphabetical order and checking for duplicate port names. Arguably these should be handled with construction or canonicalization? for the former and verification or the latter. I'm tracking the known verification stuff in #519.